### ISSUE TITLE
Undefined method "method_class" for AMQ::Protocol::BodyFrame

### ISSUE THREAD

**Issue by alexnorthsoul** [ruby-amqp/amqp#218]
Hello, guys

unfortunately, there's a bug

```
undefined method `method_class' for #<AMQ::Protocol::BodyFrame:0x0000001e8a60b0>

ruby/2.1.0/gems/amqp-1.5.0/lib/amqp/session.rb:1112:in `frameset_complete?'
ruby/2.1.0/gems/amqp-1.5.0/lib/amqp/session.rb:932:in `receive_frame'
ruby/2.1.0/gems/amqp-1.5.0/lib/amqp/session.rb:671:in `receive_data'
ruby/2.1.0/gems/eventmachine-1.0.4/lib/eventmachine.rb:187:in `run_machine
```

[Here's a line](https://github.com/ruby-amqp/amqp/blob/master/lib/amqp/session.rb#L1112), which raises an exception:

``` ruby
first_frame.final? || (first_frame.method_class.has_content? && content_complete?(frames[1..-1]))
```

in case, when **first_frame.final?** evaluates to **false**.

This happens if first_frame is a [**HeaderFrame**](https://github.com/ruby-amqp/amq-protocol/blob/master/lib/amq/protocol/frame.rb#L141) or a [**BodyFrame**](https://github.com/ruby-amqp/amq-protocol/blob/master/lib/amq/protocol/frame.rb#L187)

**Comment by Inanepenguin** [comment]
@alexnorthsoul in what context does this happen? A BodyFrame is not supposed to be the first frame in the buffer (from my understanding). I was fighting this issue myself today.

For me, an exception was thrown on the callback block of the first message received, causing the channel's buffer to not be cleared by this line:
```ruby
clear_frames_on(frame.channel) if @frames[frame.channel]
```
in `#receive_frame`. So each subsequent message received on that channel would call `first_frame.method_class` on the leftover BodyFrame.

I was about to submit an issue myself but just saw yours.

My thought is that `#receive_frameset` should have some sort of assurance of cleanup. Even if one of my messages blows out the top of the call stack, AMQP shouldn't leave itself in a bad state.

Anywhere you call user specified code without a guaranteed cleanup there can be bugs like the one outlined above.

**Comment by michaelklishin** [comment]
@Inanepenguin's hypothesis sounds realistic

**Comment by Inanepenguin** [comment]
@michaelklishin any chance the "graceful cleanup" in `#receive_format` could make it in? I can look into trying to submit a PR, I'd have to come up to speed on the style/best practice for the repo. Like I said, I was about to open an issue myself as this just happened to me today. Very coincidental timing!

**Comment by michaelklishin** [comment]
Feel free to submit a PR, thank you.

**Comment by relonger** [comment]
The issue is still exists in gem 1.6.0 but solved in 1.5.1. Do you plan to update 1.6.x gem ?

**Comment by michaelklishin** [comment]
@relonger that's curious. What specific commit is in `1.5.x-stable` but not master? Happy to release a new `1.6.x` version later today or tomorrow.

**Comment by relonger** [comment]
This commit https://github.com/ruby-amqp/amqp/commit/dd0b2de462ca37524d1a804d3210cb8e73a6c450 seems to solve the problem.

**Comment by michaelklishin** [comment]
@relonger I cherry-picked 02a92989c44bae9cc941139987710d3b0319433b to master, please give it a try.

**Comment by maattdd** [comment]
Could you release a 1.6.1 with this commit included ?

**Comment by michaelklishin** [comment]
@maattdd do you confirm that the issue is gone in master?

**Comment by maattdd** [comment]
@michaelklishin  yes I confirm this bug is in 1.6.0 but fixed in master.

**Comment by michaelklishin** [comment]
@maattdd I released `1.7.0`.
